evidence/reactor.go

```diff
@@ -21,233 +46,363 @@ const (
-func (evR *Reactor) Receive(chID byte, src p2p.Peer, msgBytes []byte) {
-	evis, err := decodeMsg(msgBytes)
-	if err != nil {
-		evR.Logger.Error("Error decoding message", "src", src, "chId", chID, "err", err)
-		evR.Switch.StopPeerForError(src, err)
-		return
+// handleMessage handles an Envelope sent from a peer on a specific p2p Channel.
+// It will handle errors and any possible panics gracefully. A caller can handle
+// any error returned by sending a PeerError on the respective channel.
+func (r *Reactor) handleMessage(chID p2p.ChannelID, envelope p2p.Envelope) (err error) {
+	defer func() {
+		if e := recover(); e != nil {
+			err = fmt.Errorf("panic in processing message: %v", e)
+		}
+	}()
+
+	switch chID {
+	case EvidenceChannel:
+		err = r.handleEvidenceMessage(envelope)
+
+	default:
+		err = fmt.Errorf("unknown channel ID (%d) for envelope (%v)", chID, envelope)
 	}
 
-	for _, ev := range evis {
-		err := evR.evpool.AddEvidence(ev)
-		switch err.(type) {
-		case *types.ErrInvalidEvidence:
-			evR.Logger.Error(err.Error())
-			// punish peer
-			evR.Switch.StopPeerForError(src, err)
+	return err
+}
+
+// processEvidenceCh implements a blocking event loop where we listen for p2p
+// Envelope messages from the evidenceCh.
+func (r *Reactor) processEvidenceCh() {
+	defer r.evidenceCh.Close()
+
+	for {
+		select {
+		case envelope := <-r.evidenceCh.In():
+			if err := r.handleMessage(r.evidenceCh.ID(), envelope); err != nil {
+				r.Logger.Error("failed to process message", "ch_id", r.evidenceCh.ID(), "envelope", envelope, "err", err)
+				r.evidenceCh.Error() <- p2p.PeerError{
+					PeerID:   envelope.From,
+					Err:      err,
+					Severity: p2p.PeerErrorSeverityLow,
+				}
+			}
+
+		case <-r.closeCh:
+			r.Logger.Debug("stopped listening on evidence channel; closing...")
 			return
-		case nil:
-		default:
-			// continue to the next piece of evidence
-			evR.Logger.Error("Evidence has not been added", "evidence", evis, "err", err)
 		}
 	}
 }
 
-// SetEventBus implements events.Eventable.
-func (evR *Reactor) SetEventBus(b *types.EventBus) {
-	evR.eventBus = b
+// processPeerUpdate processes a PeerUpdate, returning an error upon failing to
+// handle the PeerUpdate or if a panic is recovered. For new or live peers it
+// will check if an evidence broadcasting goroutine needs to be started. For
+// down or removed peers, it will check if an evidence broadcasting goroutine
+// exists and signal that it should exit.
+//
+// FIXME: The peer may be behind in which case it would simply ignore the
+// evidence and treat it as invalid. This would cause the peer to disconnect.
+// The peer may also receive the same piece of evidence multiple times if it
+// connects/disconnects frequently from the broadcasting peer(s).
+//
+// REF: https://github.com/tendermint/tendermint/issues/4727
+func (r *Reactor) processPeerUpdate(peerUpdate p2p.PeerUpdate) (err error) {
+	defer func() {
+		if e := recover(); e != nil {
```

Comment:
That's true. Although, I think there is no harm in doing so. Why not protect against potentially faulty code in peer update handling?

But then again, now that I think about it, maybe we want to the panic to occur to point out the bug such that it can be fixed.

So yeah, I think it makes sense to remove this.

Reply:
Yeah, but that sort of goes for all code, why not put panic recovery everywhere? 🙂 I don't know, no harm in keeping it as you say, just trying to reduce boilerplate. Your call.

Reply:
I'd like to remove it because we could mask flaws. I'll handle the evidence reactor in this PR, but other reactors in another PR.